src/main/java/org/oscii/morph/RuleScored.java

```diff
@@ -0,0 +1,185 @@
+        filterTransformations(embeddings, params);
+        return hitRate;
+    }
+
+    private void filterTransformations(EmbeddingContainer embeddings, ScoringParams params) {
+        topDirections = new ArrayList<>();
+        transformations = new ArrayList<>();
+        HashMap<RulePair, List<Transformation>> remaining = new HashMap<>(hits);
+        int previousSize = -1;
+        while (topDirections.size() > previousSize) {
+            previousSize = topDirections.size();
+            // Find the direction that explains the most pairs
+            remaining.values().stream()
+                    .flatMap(Collection::stream)
+                    .collect(groupingBy(t -> t.direction))
+                    .entrySet().stream()
+                    .max((s, t) -> Long.compare(s.getValue().size(), t.getValue().size()))
+                    .ifPresent(mostCommon -> {
+                        int size = mostCommon.getValue().size();
+                        log.debug("  #{} direction has {} hits", topDirections.size() + 1, size);
+                        if (size >= params.minSizeDirection) {
+                            topDirections.add(mostCommon.getKey());
+                            // Remove all pairs explained by this direction
+                            mostCommon.getValue().forEach(t -> remaining.remove(t.rule));
+                        }
+                    });
+        }
+
+        // Find all valid transformations for all embedded rules
+        log.debug("  Finding transformations for {} directions", topDirections.size());
+        topDirections.forEach(d -> embedded.forEach(r -> {
+            Transformation t = scorePairAndDirection(r.pair, d, embeddings, params.maxRankTransformation);
+            if (t.rank <= params.maxRankTransformation && t.cosine >= params.minCosineTransformation) {
+                transformations.add(t);
+            }
+        }));
+        transformations.sort((t, u) -> Double.compare(t.rank - t.cosine, u.rank - u.cosine));
+        log.debug("  {} transformations found", transformations.size());
+    }
+
+    public List<Transformation> getTransformations() {
+        return transformations;
+    }
+
+    private boolean embedded(RuleLexicalized r, EmbeddingContainer embeddings) {
+        for (String w : Arrays.asList(new String[]{r.pair.input, r.pair.output})) {
+            if (!embeddings.contains(w)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void add(float[] x, float[] y, float[] z) {
+        for (int i = 0; i < z.length; i++) {
+            z[i] = x[i] + y[i];
+        }
+    }
+
+    private Transformation scorePairAndDirection(RulePair r, RulePair d, EmbeddingContainer vs, int rankThreshold) {
+        add(d.getDirection(vs), vs.getRawVector(r.input), added);
+        double cosine = VectorMath.cosineSimilarity(d.getDirection(vs), r.getDirection(vs));
+        List<String> neighbors = vs.neighbors(added, rankThreshold);
```

Comment:
I'm pretty sure that this is not how Radu implemented it. The paper says that they are using the "cosine-similarity rank function". But as cosine similarity only measures the angle between two vectors I like your solution better.